cabby/rvs/templates.py

```diff
@@ -0,0 +1,198 @@
+      'V1'), Nonterminal('NO_GOAL'))),
+    # The goal in the begining and the rest of the instruction body assuming
+    # goal already mentioned.
+    Production(Nonterminal('S'), (Nonterminal('V1_GOAL'),
+                    Nonterminal('WITH_GOAL'))),
+    # A verb and 'to the' and then goal mention and the rest of the instruction
+    # body.
+    Production(Nonterminal('V1_GOAL'), (Nonterminal('V1'),
+                      Nonterminal('V1_CON'))),
+    # A goal mention and the rest of the instruction body.
+    Production(Nonterminal('WITH_GOAL'), (Nonterminal('GOAL'),
+                        Nonterminal('M_G'))),
+    # Main part of the instruction without verb in begining and resuming
+    # sentence.
+    Production(Nonterminal('M_G'), (Nonterminal('MAIN_NO_V'),
+                    Nonterminal('E'))),
+    # Main part of the instruction, adding a new sentence.
+    Production(Nonterminal('M_G'), (Nonterminal('MAIN'),
+                    Nonterminal('E'))),
+    # End part - (1) near pivot assuming goal already mentioned; and (2) avoid
+    # sentence.
+    Production(Nonterminal('E'), (Nonterminal('NEAR_GOAL_START'),
+                    Nonterminal('AVOID'))),
+    # End part - (1) near pivot assuming goal not mentioned yet; and (2) avoid
+    # sentence.
+    Production(Nonterminal('E'), (Nonterminal('NEAR_GOAL_END'),
+                    Nonterminal('AVOID'))),
+    # Main part of the instruction without verb in begining and resuming
+    # sentence assuming no goal mentioned before.
+    Production(Nonterminal('NO_GOAL'), (Nonterminal('MAIN_NO_V'),
+                      Nonterminal('G'))),
+    # Add Goal to main part and then resume instruction by adding an
+    # ending(near+avoid).
+    Production(Nonterminal('G'), (Nonterminal('GOAL_END'),
+                    Nonterminal('E'))),
+    # Add Goal with near and then add an avoid sentenece.
+    Production(Nonterminal('G'), (Nonterminal('NEAR_GOAL_END'),
+                    Nonterminal('AVOID'))),
+    # Termial for IN+DT after verb.
+    Production(Nonterminal('V1_CON'), ('to the',)),
+
+  ]
+
+  prods += add_rules('V2', V2)
+  prods += add_rules('AVOID', AVOID)
+  prods += add_rules('NEAR_GOAL_START', NEAR_GOAL_START)
+  prods += add_rules('NEAR_GOAL_END', NEAR_GOAL_END)
+  prods += add_rules('GOAL', GOAL)
+  prods += add_rules('GOAL_END', GOAL_END)
+  prods += add_rules('MAIN_NO_V', MAIN_NO_V)
+  prods += add_rules('MAIN', MAIN)
+  prods += add_rules('V1', V1)
+
+  grammar = CFG(Nonterminal('S'), prods)
+
+  # Generate templates.
+  templates = []
+  for sentence in generate(grammar):
+
+    sentence = ' '.join(sentence)
+
+    if sentence[-1] != '.':
+      sentence += '.'
+    sentence = sentence.replace(" .", ".")
+    sentence = sentence.replace(" ,", ",")
+    sentence = sentence.replace("..", ".")
+    sentence = sentence.replace("   ", " ")
+    sentence = sentence.replace("  ", " ")
+    templates.append(sentence)
+
+  templates_df = pd.DataFrame(
```

Comment:
I can't reply to your response to my question earlier, so bringing the context back to here:

- Me: Why use a data frame here instead of defining a data structure to hold these various elements and passing the templates around a sequence of Template items?

- You: easier to condition on.
E.g,
current_templates = gen_templates[
current_templates['beyond_pivot'] == False]

My problem is that dataframes are hard to reason about here, because there is no good predefined structure with functions governing what can be done with them. But it's okay for now -- it's working so let's go with it as is for the time being.

You can resolve this comment -- no action necessary other than to read it!